### DISCUSSION TITLE
Dynamic DAG generation using data from external data store

### DISCUSSION THREAD

**Discussion by chnaveen138** [apache/airflow#34798] [General]
Hi, 
We are using MWAA in AWS for hosting our airflow DAGs. We have list of workflows with different names and schedules. This list will be updated by external users, so it can increase time to time. Now we want to have a separate Airflow DAG for each workflow with desired schedule based on the list. We're thinking to have a python file like `dag_generator.py`  that can create dags based on the list. Since this list is dynamic, we couldn't find a way to do that as we will need to fetch this list from external data store and we cannot do that during parsing, as it doesn't follow airflow best practices, as parsing would be a frequent operation. Is there a neat and any best approach to solve this?

Please suggest any approach other than creating new file for each DAG based on user's new entry into the list as we don't want to go in this way. 

Thanks in Advance.

**Comment by Taragolis** [comment]
In case if you use non-Managed service, I would recommend to attach AWS EFS (NFS share) and store metadata for DAG generation here, e.g. some YAML/JSON and etc, which might be used for generate dynamic DAGs.

In your case I think you have only one option to make it more quicker in execution by use this approach [Optimizing DAG parsing delays during execution](https://airflow.apache.org/docs/apache-airflow/stable/howto/dynamic-dag-generation.html#optimizing-dag-parsing-delays-during-execution)

Maybe when AWS have Airflow 2.7 on their MWAA you might also use Variables with [enabling the experimental cache](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#config-secrets-use-cache), 2.7.0+ (experimental, so it could be removed in any moment), it is not silver bullet and also might have some side effect, becase it in general new feature

From some discussion in the slack about caching in Variables/Connections:
> In the first glance Variable caching (experimental, Airflow 2.7.0+) might help resolve this issues, BTW huge bills for MWAA Users it is main reason why this feature added in Airflow, however it is a new feature, it is disabled by default and not a lot of statistics about this feature yet.  For me it would be interesting how it deals with HA schedulers.

Regardless of selected approach make sure you follow [Dynamic DAG Generation](https://airflow.apache.org/docs/apache-airflow/stable/howto/dynamic-dag-generation.html) and [Fine-tuning your Scheduler performance](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/scheduler.html#fine-tuning-your-scheduler-performance) as much as possible, e.g. if your DAGs generated from single file then you basically lost a parallel DAG Processing and  this might leads additional issues, e.g. timeouts during DAG generation
